aten/src/ATen/native/Distance.cpp

```diff
@@ -75,4 +88,269 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
   return w12.div_((w1 * w2).clamp_min_(eps));
 }
 
+template <typename scalar_t>
+static void to_squareform_frame(
+    scalar_t *input_p, scalar_t *output_p, int64_t dimension, int64_t size)
+{
+  for (int64_t i = 0, idx = 0; i < dimension; i++)
+  {
+    for (int64_t j = i + 1; j < dimension && idx < size; j++)
+    {
+      scalar_t *dest_ij = output_p + i * dimension + j;
+      scalar_t *dest_ji = output_p + j * dimension + i;
+      *dest_ij = input_p[idx];
+      *dest_ji = input_p[idx];
+      idx++;
+    }
+  }
+}
+
+template <typename scalar_t>
+static void from_squareform_frame(
+    scalar_t *input_p, scalar_t *output_p, int64_t dimension, int64_t size)
+{
+  for (int64_t i = 0, idx = 0; i < dimension; i++)
+  {
+    for (int64_t j = i + 1; j < dimension && idx < size; j++)
+    {
+      output_p[idx] = *(input_p + i * dimension + j);
+      idx++;
+    }
+  }
+}
+
+template <typename scalar_t>
+static void to_squareform_backward_frame(
+    scalar_t *input_p, scalar_t *output_p, int64_t dimension, int64_t size)
+{
+  for (int64_t i = 0, idx = 0; i < dimension; i++)
+  {
+    for (int64_t j = i + 1; j < dimension && idx < size; j++)
+    {
+      output_p[idx] = *(input_p + i * dimension + j);
+      output_p[idx] += *(input_p + j * dimension + i);
+      idx++;
+    }
+  }
+}
+
+template <typename scalar_t>
+static void from_squareform_backward_frame(
+    scalar_t *input_p, scalar_t *output_p, int64_t dimension, int64_t size)
+{
+  for (int64_t i = 0, idx = 0; i < dimension; i++)
+  {
+    for (int64_t j = i + 1; j < dimension && idx < size; j++)
+    {
+      scalar_t *dest_ij = output_p + i * dimension + j;
+      *dest_ij = input_p[idx];
+      idx++;
+    }
+  }
+}
+
+Tensor to_squareform_cpu(const Tensor& self)
+{
+  AT_CHECK(self.dim() == 1,
+    "Expected 1-dimensional (vector-form) distance,",
+    " but got tensor of dimensionality ", self.dim());
+  int64_t size = self.size(0);
+  int64_t d = getSquareFormDimension(size);
+  Tensor out_tensor = at::zeros({d, d}, self.options());
+  AT_DISPATCH_ALL_TYPES(self.type(), "squareform", [&] {
+      to_squareform_frame<scalar_t> (self.data<scalar_t>(),
+        out_tensor.data<scalar_t>(), d, size);
+      }
+      );
+  return out_tensor;
+}
+
+Tensor from_squareform_cpu(const Tensor& self)
+{
+  AT_CHECK(self.dim() == 2,
+    "Expected 2-dimensional (square-form) distance,",
+    " but got tensor of dimensionality ", self.dim());
+  AT_CHECK(self.size(0) == self.size(1),
+    "expected square-form distance as input, size(0) ",
```

Comment:
this isn't a particularly helpful error message; the user might not know what a square-form distance is.  What about just saying square matrix?